### ISSUE TITLE
Problem with mod_h2 and csrf tokens in fcgi apps

### ISSUE THREAD

**Issue by jfmcbrayer** [icing/mod_h2#13]
I'm sorry if this bug report is not very good, as I don't really know what's going on.

I'm unable to log in to a number of PHP apps I'm running as fastcgi after enabling mod_h2. Apps in question are [rainloop mail](http://rainloop.net/) and roundcube mail. I'm also having trouble with ownCloud, but it's running on another server and behind a mod_proxy setup on this server, so it may be a different issue.

With rainloop: login page loads fine. Try to log in (it makes an ajax POST and gets json back), it pops up an error saying 'invalid token'. There's nothing in any of the server logs. Tested in Firefox and Chrome on Windows (server is Fedora 20: Apache/2.4.10 (Fedora) OpenSSL/1.0.1e-fips mod_fcgid/2.3.9 PHP/5.5.23 mod_wsgi/3.5 Python/2.7.5), 

Basically the same thing happens with roundcube mail, except that the error is "Invalid request! No data was saved."

Disable the loading of mod_h2, and the same server config works. I'm assuming this is a CSRF token that's failing; I have no idea why this is breaking except that maybe things are loading in a different order?

**Comment by icing** [comment]
Ok, seems like i need to include php in my sandbox tests. 

I have to get back to you next week on this. Thanks for your patience.

cheers,

  Stefan

> Am 17.04.2015 um 16:01 schrieb jfmcbrayer <notifications@github.com>:
> 
> I'm sorry if this bug report is not very good, as I don't really know what's going on.
> 
> I'm unable to log in to a number of PHP apps I'm running as fastcgi after enabling mod_h2. Apps in question are rainloop mail and roundcube mail. I'm also having trouble with ownCloud, but it's running on another server and behind a mod_proxy setup on this server, so it may be a different issue.
> 
> With rainloop: login page loads fine. Try to log in (it makes an ajax POST and gets json back), it pops up an error saying 'invalid token'. There's nothing in any of the server logs. Tested in Firefox and Chrome on Windows (server is Fedora 20: Apache/2.4.10 (Fedora) OpenSSL/1.0.1e-fips mod_fcgid/2.3.9 PHP/5.5.23 mod_wsgi/3.5 Python/2.7.5),
> 
> Basically the same thing happens with roundcube mail, except that the error is "Invalid request! No data was saved."
> 
> Disable the loading of mod_h2, and the same server config works. I'm assuming this is a CSRF token that's failing; I have no idea why this is breaking except that maybe things are loading in a different order?
> 
> —
> Reply to this email directly or view it on GitHub.
> 

<green/>bytes GmbH
Hafenweg 16, 48155 Münster, Germany
Phone: +49 251 2807760. Amtsgericht Münster: HRB5782

**Comment by jfmcbrayer** [comment]
With beb622311023dcb097c839d2d61e39a1c9de9878 it's much better, but not perfect. I can log in, and then get a 'invalid token' error in the area where the message list is supposed to load. Clicking on a folder now gets me a list of messages, but subsequently, various requests may or may not work.

**Comment by icing** [comment]
This webapp seems to cache heavily. When I tested today initially, I saw some pages with "invalid token", but the I force-reloaded my browser and all seems fine now. Browsing large imap folders...

Could you try this on your installation? Clearing browser cache (don't know how to do that best under Windows...). Thanks!

**Comment by icing** [comment]
Hold the horses! I found a bug in my handling of requests without content-length. Will update when a fix is pushed...

**Comment by icing** [comment]
Please retest with v0.5.2. It is passing all my tests now and looks fine on my rainloop.

**Comment by jfmcbrayer** [comment]
Will test asap.

**Comment by jfmcbrayer** [comment]
Rainloop is working about 95% now. Opening some mailboxes will hang for a while and then return a 'Can't get message list' error. For some reason, threads can't be expanded. If you can suggest a CustomLog line to get the info you need on these requests, I'll do more testing.

Actually, a little more testing shows everything works right if message threading is turned off, which is weird.

**Comment by icing** [comment]
request timeouts happen with threaded display. I will have a look how that happens. 

//stefan

**Comment by icing** [comment]
Could you please check the current github master? I removed the default timeout handlers. Thanks.

**Comment by jfmcbrayer** [comment]
It's still hanging on loading certain mailboxes. I'll try to post logs later, but I can't do it from where I am. Basically what I'm seeing is page after page of h2_mplx(133) trywait on data for 200.000000 ms.

**Comment by icing** [comment]
Ok, I think I need to adjust some log levels for it to make sense. This endless waiting on data from requests means that some request processing hangs or is not recognizing end of streams correctly. 

I will let you know when there is a version with more proper logging available.

//Stefan

> Am 23.04.2015 um 13:49 schrieb jfmcbrayer <notifications@github.com>:
> 
> It's still hanging on loading certain mailboxes. I'll try to post logs later, but I can't do it from where I am. Basically what I'm seeing is page after page of h2_mplx(133) trywait on data for 200.000000 ms.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

<green/>bytes GmbH
Hafenweg 16, 48155 Münster, Germany
Phone: +49 251 2807760. Amtsgericht Münster: HRB5782

**Comment by jfmcbrayer** [comment]
One problem seems to be a bad rainloop update, not mod_h2: expanding collapsed threads is simply broken in the latest rainloop, even without mod_h2.

Still can't log into roundcubemail with mod_h2, but that may be a little trickier for you to test (it's a bigger package than rainloop).

**Comment by icing** [comment]
Closing this issue then, as it seems to be solved by v0.5.2
